### DISCUSSION TITLE
0.1.0

### DISCUSSION THREAD

**Discussion by ramsey** [php-identifier/identifier#16] [General]
Initial publication of specification draft and interfaces: https://php-identifier.readthedocs.io/en/0.1.0/

<hr /><em>This discussion was created from the release <a href='https://github.com/php-identifier/identifier/releases/tag/0.1.0'>0.1.0</a>.</em>

**Comment by dgoosens** [comment]
hi @ramsey 

That's already quite a nice "first draft"... Awesome!  

Very personal preference and probably very irrelevant, the only thing I do differently currently is use the term `generate` instead of `create` for the `IdentifierFactory`  
I chose that term after reading ao this <https://ell.stackexchange.com/questions/76985/create-vs-generate-vs-make> and this <https://hinative.com/questions/2219> (very old posts)  
Also, I prefer `Generator` objects over `Creator`objects...

**Reply by ramsey** [reply]
I chose "create" instead of "generate" or "make" because "create" is used in almost all factory design patterns examples and implementations I could find. Factories build or create things from raw materials. They don't generate things.

PHP already has a concept of generators, and while you could argue that IDs are generated rather than created, I don't want to cause any unnecessary confusion, since these classes are factories and do not provide any iteration functionality. (IMO, the IDs themselves are not *generated*. Rather, the data used to create the IDs is what is generated, and that's not a concern of these interfaces.)

**Comment by GromNaN** [comment]
Hello @ramsey, this is a very clean start.

For now, I don't get the actual use-cases of this interfaces. Do you have actual examples where they are needed?

**Reply by ramsey** [reply]
The primary use-case is to decouple ID creation from the implementation by creating a standard for applications and libraries.

Consider something like this contrived example:

```php
use Ramsey\Uuid\Uuid;
use Ramsey\Uuid\UuidInterface;

class Foo
{
    public readonly UuidInterface $id;

    public function __construct()
    {
        $id = Uuid::uuid7();
    }
}
```

The application where `Foo` is defined is bound to the ramsey/uuid library. This is a tradeoff that some might consider okay, but this dependency creates limitations that affect the maintainability of the code. I can't easily swap out the ID creation with ULIDs, for example, and if I did, my application might have a lot of assumptions about the IDs, so it might expect UUIDs (not just ramsey/uuid) everywhere.

If we introduce a standard interface for identifiers, then the IDs become opaque to the application, and we can easily change the strategy for ID creation.

```php
use Identifier\Identifier;
use Identifier\IdentifierFactory;

class Foo
{
    public readonly Identifier $id;

    public function __construct(IdentifierFactory $factory)
    {
        $id = $factory->create();
    }
}
```

I can take this further in my application by setting up a service locator / dependency-injection container and configuring it to use any kind of identifier my application needs:

```php
// This is pseudo-code to show the concept. A specific DI package is not implied.
$container->add(IdentifierFactory::class, new Ramsey\Identifier\UuidFactory());
$container->add(Foo::class)->args(IdentifierFactory::class);
```

It's possible to do this with the current state of affairs, but the types are still specific to libraries (i.e., ramsey/uuid, symfony/uid, robinvdvleuten/ulid, etc.), which binds your application to those libraries, making it more difficult to switch between them.

Additionally, if any third-party libraries want to create IDs on behalf of your application, they must either implement their own adapters for different ID libraries, or they must require your application to support a specific ID library. By creating this abstraction, third-party libraries can create IDs on behalf of your application without being bound to a specific ID creation implementation.

**Comment by lstrojny** [comment]
Hey Ben,
looks good! One question though: since `IntegerIdentifier` is limited to INT_MAX/INT_MIN, how would one represent a big int, except using a `StringIdentifier`?

**Reply by ramsey** [reply]
I have this under the `IntegerIdentifier` section:

> Implementations MAY support identifiers greater than `PHP_INT_MAX` and less than `PHP_INT_MIN`. In this case, the `toInteger()` method SHOULD return a string value. If the return value is a string, it MUST be a numeric string.

And this is the signature of `toInteger()`:

```php
public function toInteger(): int | string;
```

Let me know how I can make it more clear that big integers are supported.

**Reply by lstrojny** [reply]
Hah, looks like I didn't read it carefully enough. That makes a lot of sense, thanks.

**Reply by ramsey** [reply]
I might need to update the `IntegerIdentifierFactory` section to say something similar for the `createFromInteger()` method.

**Reply by lstrojny** [reply]
I wonder if it would make sense to always represent it as a string to simplify types
